File: src/main/java/io/github/bot/terminal/application/domains/common/action/entity/SqlSelectAsTextAction.java
package io.github.bot.terminal.application.domains.common.action.entity;


import io.github.bot.terminal.application.domains.db_connection.entity.DbConnection;
import io.github.bot.terminal.application.domains.notificarion_api.entity.Message;
import io.github.bot.terminal.application.domains.notificarion_api.entity.NotificationApi;
import lombok.Data;
import lombok.experimental.Accessors;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.text.StringSubstitutor;
import org.springframework.jdbc.core.JdbcTemplate;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Slf4j
@Data
@Accessors(chain = true)
public class SqlSelectAsTextAction implements Action {

    public SqlSelectAsTextAction(String select, String resultTemplate, DbConnection dbConnection) {
        this.select = select;
        this.resultTemplate = resultTemplate;
        this.dbConnection = dbConnection;
    }

    private final String select;

    private final DbConnection dbConnection;

    private final String resultTemplate;

    @Override
    public ActionResult execute() {
        try {
            JdbcTemplate template = dbConnection.createJdbcTemplate();
            List<Map<String, Object>> rows = template.queryForList(select);
            StringBuilder responseBuilder = new StringBuilder();
            for (Map<String, Object> row : rows) {
                responseBuilder.append(renderRow(row));
                responseBuilder.append(System.lineSeparator());
            }
            String responseMessage = responseBuilder.toString();
            return new ActionResultImpl(responseMessage);
        } catch (Exception ex) {
            log.warn("Problem to execute select: {}", select);
            throw ex;
        }
    }

    private String renderRow(Map<String, Object> row) {
        Map model = new HashMap();
        row.keySet().forEach(k -> {
            model.put(k, row.get(k));
            model.put(k.toLowerCase(), row.get(k));
        });
        StringSubstitutor sub = new StringSubstitutor(model);
        return sub.replace(resultTemplate);
    }

    public class ActionResultImpl implements ActionResult<String> {

        private final String result;

        public ActionResultImpl(String result) {
            this.result = result;
        }

        @Override
        public void notify(NotificationApi notificationApi) {
            notificationApi.sendMessage(new Message(result));
        }

        @Override
        public String getResult() {
            return result;
        }

    }
}

File: src/main/kotlin/io/github/bot/terminal/application/domains/common/action/entity/SqlSelectAsTextAction.kt
package io.github.bot.terminal.application.domains.common.action.entity

import io.github.bot.terminal.application.domains.db_connection.entity.DbConnection
import io.github.bot.terminal.application.domains.notificarion_api.entity.Message
import io.github.bot.terminal.application.domains.notificarion_api.entity.NotificationApi
import org.apache.commons.text.StringSubstitutor
import java.util.*
import java.util.function.Consumer
import kotlin.collections.HashMap

class SqlSelectAsTextAction(private val select: String,
                            private val resultTemplate: String,
                            private val dbConnection: DbConnection<*>) : Action {

    override fun execute(): ActionResult<*> {
        return try {
            val template = dbConnection.createJdbcTemplate()
            val rows = template.queryForList(select)
            val responseBuilder = StringBuilder()
            for (row in rows) {
                responseBuilder.append(renderRow(row))
                responseBuilder.append(System.lineSeparator())
            }
            val responseMessage = responseBuilder.toString()
            ActionResultImpl(responseMessage)
        } catch (ex: Exception) {
            throw ex
        }
    }

    private fun renderRow(row: Map<String, Any>): String {
        val model: MutableMap<String, Any?> = HashMap()
        row.keys.forEach(Consumer { k: String ->
            model[k] = row[k]
            model[k.toLowerCase()] = row[k]
        })
        val sub = StringSubstitutor(model)
        return sub.replace(resultTemplate)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as SqlSelectAsTextAction

        if (select != other.select) return false
        if (resultTemplate != other.resultTemplate) return false
        if (dbConnection != other.dbConnection) return false

        return true
    }

    override fun hashCode(): Int {
        var result = select.hashCode()
        result = 31 * result + resultTemplate.hashCode()
        result = 31 * result + dbConnection.hashCode()
        return result
    }

    inner class ActionResultImpl(override val result: String?) : ActionResult<String?> {
        override fun notify(notificationApi: NotificationApi) {
            if(result != null){
                notificationApi.sendMessage(Message(result))
            }
        }
    }

}

